From 9e3a25ce8d15adb66ec8cee18f539c474a6526e2 Mon Sep 17 00:00:00 2001
From: Ioannis Galanommatis <galanom@gmail.com>
Date: Fri, 15 Dec 2017 23:59:41 +0200
Subject: [PATCH] allow pools with shared chunks

---
 include/linux/genalloc.h | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/include/linux/genalloc.h b/include/linux/genalloc.h
index 78eef3ef1785..75d348a04445 100644
--- a/include/linux/genalloc.h
+++ b/include/linux/genalloc.h
@@ -98,6 +98,9 @@ struct genpool_data_fixed {
 };
 
 extern struct gen_pool *gen_pool_create(int, int);
+extern struct gen_pool_chunk *gen_pool_create_chunk(
+	unsigned long virt, phys_addr_t phys, size_t size, int nid);
+extern void gen_pool_add_chunk(struct gen_pool *pool, struct gen_pool_chunk *chunk);
 extern phys_addr_t gen_pool_virt_to_phys(struct gen_pool *pool, unsigned long);
 extern int gen_pool_add_virt(struct gen_pool *, unsigned long, phys_addr_t,
 			     size_t, int);
-- 
2.13.0

